alembic/script/revision.py

```diff
@@ -886,6 +895,122 @@ def _iterate_related_revisions(self, fn, targets, map_, check=False):
+            if child.dependencies is not None
+            for parent in util.to_tuple(child.dependencies)
+        ]
+        return sqlautil.topological.sort(edges, allitems)
+
+    def walk_down(self, start, steps, label):
+        """ Walk down the tree along a single path. """
+        assert steps <= 0
+        if isinstance(start, compat.string_types):
+            start = self.get_revision(start)
+        if steps == 0:
+            return start
+        if start is None:
+            raise ValueError("Can't walk back past base.")
+        children = self.get_revisions(start.down_revision)
+        if len(children) == 0:
+            raise util.CommandError(
+                f"Relative revision {label} didn't produce "
+                f"{abs(int(label))} migrations"
+            )
+        assert len(children) == 1
+        return self.walk_down(next(iter(children)), steps + 1, label)
+
+    def walk(self, start, steps, branch_label):
+        """ Walk up the tree along a single path. """
+        assert steps >= 0
+        if isinstance(start, compat.string_types):
+            start = self.get_revision(start)
+        if steps == 0:
+            return start
+        if start is None:
+            roots = [
+                rev
+                for rev in self._revision_map.values()
+                if rev is not None and rev.down_revision is None
+            ]
+            assert len(roots) == 1
+            return self.walk(roots[0], steps - 1, branch_label)
+        children = [
+            rev
+            for rev in self.get_revisions(start.nextrev)
+            if branch_label is None or branch_label in rev.branch_labels
+        ]
+        assert len(children) == 1
+        return self.walk(next(iter(children)), steps - 1, branch_label)
+
+    def _drop_inclusive(self, branch_revision, upper):
+        # Aim then is to drop :branch_revision; to do so we also need
+        # to drop its descendents and anything dependent on it.
+        drop_revisions = set(
+            self._get_descendant_nodes(
+                branch_revision,
+                include_dependencies=True,
+                omit_immediate_dependencies=False,
+            )
+        )
+        # Set logic/walking full tree might get expensive?
+        active_revisions = set(
+            self._get_ancestor_nodes(
+                self.get_revisions(upper), include_dependencies=True
+            )
+        )
+
+        # Emit revisions to drop in reverse topological sorted order.
+        drop_revisions = drop_revisions.intersection(active_revisions)
+
+        if len(drop_revisions) == 0:
+            # Empty intersection: target revs are not present.
+            raise RangeNotAncestorError(None, upper)
+
+        drop_reverse_topo_sorted = list(
+            reversed(list(self.topological_sort(drop_revisions)))
+        )
+        yield from self.get_revisions(drop_reverse_topo_sorted)
+
+    def _parse_downgrade(self, upper, lower):
```

Comment:
**mike bayer** (zzzeek) wrote:

is this doing what _relative_iterate does?  can these be unified?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/2530

Reply:
**Simon Bowly** (simonbowly) wrote:

Yes I think it is doing something similar. I'm struggling to follow _relative_iterate though; I think it's cleaner to avoid having two separate cases for relative and absolute revisions. Once you determine what's being pointed to (either using absolute or relative syntax) the logic for determining revisions to drop shouldn't depend on how the target was specified. Should be able to merge the two.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/2530